File: internal/batch/batch_test.go
package batch

import (
	"context"
	"fmt"
	"testing"
)

type mockGit struct {
	branches      []string
	merges        []string
	pushed        []string
	ffRef         string
	deleted       []string
	conflictOn    string // HeadRef that will conflict
	failOn        string // method name to fail on
}

func (m *mockGit) CreateBranchFromRef(_ context.Context, branch string, _ string) error {
	if m.failOn == "CreateBranchFromRef" {
		return fmt.Errorf("mock error")
	}
	m.branches = append(m.branches, branch)
	return nil
}

func (m *mockGit) MergeBranch(_ context.Context, _ string, sourceBranch string, _ string) (bool, error) {
	if m.failOn == "MergeBranch" {
		return false, fmt.Errorf("mock error")
	}
	if sourceBranch == m.conflictOn {
		return false, nil
	}
	m.merges = append(m.merges, sourceBranch)
	return true, nil
}

func (m *mockGit) PushBranch(_ context.Context, branch string) error {
	if m.failOn == "PushBranch" {
		return fmt.Errorf("mock error")
	}
	m.pushed = append(m.pushed, branch)
	return nil
}

func (m *mockGit) FastForwardMain(_ context.Context, ref string) error {
	if m.failOn == "FastForwardMain" {
		return fmt.Errorf("mock error")
	}
	m.ffRef = ref
	return nil
}

func (m *mockGit) DeleteBranch(_ context.Context, branch string) error {
	if m.failOn == "DeleteBranch" {
		return fmt.Errorf("mock error")
	}
	m.deleted = append(m.deleted, branch)
	return nil
}

func nopLog(string, ...any) {}

func TestCreateAndMerge_AllSuccess(t *testing.T) {
	git := &mockGit{}
	b := New(git, false, nopLog)
	prs := []PR{
		{Number: 1, HeadRef: "feature-a", Title: "Add feature A"},
		{Number: 2, HeadRef: "feature-b", Title: "Add feature B"},
	}

	result, err := b.CreateAndMerge(context.Background(), "test-1", prs)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if result.Branch != "merge-queue/batch-test-1" {
		t.Errorf("branch = %q, want %q", result.Branch, "merge-queue/batch-test-1")
	}
	if len(result.Merged) != 2 {
		t.Errorf("merged %d PRs, want 2", len(result.Merged))
	}
	if len(result.Conflicted) != 0 {
		t.Errorf("conflicted %d PRs, want 0", len(result.Conflicted))
	}
	if len(git.branches) != 1 || git.branches[0] != "merge-queue/batch-test-1" {
		t.Errorf("branches = %v", git.branches)
	}
	if len(git.merges) != 2 {
		t.Errorf("merges = %v", git.merges)
	}
	if len(git.pushed) != 1 {
		t.Errorf("pushed = %v", git.pushed)
	}
}

func TestCreateAndMerge_WithConflict(t *testing.T) {
	git := &mockGit{conflictOn: "feature-b"}
	b := New(git, false, nopLog)
	prs := []PR{
		{Number: 1, HeadRef: "feature-a", Title: "A"},
		{Number: 2, HeadRef: "feature-b", Title: "B"},
		{Number: 3, HeadRef: "feature-c", Title: "C"},
	}

	result, err := b.CreateAndMerge(context.Background(), "test-2", prs)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(result.Merged) != 2 {
		t.Errorf("merged %d PRs, want 2", len(result.Merged))
	}
	if len(result.Conflicted) != 1 {
		t.Fatalf("conflicted %d PRs, want 1", len(result.Conflicted))
	}
	if result.Conflicted[0].Number != 2 {
		t.Errorf("conflicted PR = #%d, want #2", result.Conflicted[0].Number)
	}
}

func TestCreateAndMerge_DryRun(t *testing.T) {
	git := &mockGit{}
	b := New(git, true, nopLog)
	prs := []PR{{Number: 1, HeadRef: "feature-a", Title: "A"}}

	result, err := b.CreateAndMerge(context.Background(), "dry", prs)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(result.Merged) != 1 {
		t.Errorf("expected 1 merged in dry run")
	}
	if len(git.branches) != 0 {
		t.Errorf("dry run should not create branches")
	}
	if len(git.pushed) != 0 {
		t.Errorf("dry run should not push")
	}
}

func TestCreateAndMerge_CreateBranchError(t *testing.T) {
	git := &mockGit{failOn: "CreateBranchFromRef"}
	b := New(git, false, nopLog)
	_, err := b.CreateAndMerge(context.Background(), "err", []PR{{Number: 1, HeadRef: "f", Title: "T"}})
	if err == nil {
		t.Fatal("expected error")
	}
}

func TestCreateAndMerge_NoPRs(t *testing.T) {
	git := &mockGit{}
	b := New(git, false, nopLog)
	result, err := b.CreateAndMerge(context.Background(), "empty", nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(result.Merged) != 0 {
		t.Errorf("expected no merged PRs")
	}
	// Should still create the branch but not push (no merged PRs)
	if len(git.pushed) != 0 {
		t.Errorf("should not push with no merged PRs")
	}
}

func TestCompleteMerge(t *testing.T) {
	git := &mockGit{}
	b := New(git, false, nopLog)
	err := b.CompleteMerge(context.Background(), "merge-queue/batch-1")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if git.ffRef != "merge-queue/batch-1" {
		t.Errorf("ffRef = %q", git.ffRef)
	}
	if len(git.deleted) != 1 || git.deleted[0] != "merge-queue/batch-1" {
		t.Errorf("deleted = %v", git.deleted)
	}
}

func TestCompleteMerge_DryRun(t *testing.T) {
	git := &mockGit{}
	b := New(git, true, nopLog)
	err := b.CompleteMerge(context.Background(), "merge-queue/batch-1")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if git.ffRef != "" {
		t.Errorf("dry run should not fast-forward")
	}
}

func TestCompleteMerge_FFError(t *testing.T) {
	git := &mockGit{failOn: "FastForwardMain"}
	b := New(git, false, nopLog)
	err := b.CompleteMerge(context.Background(), "merge-queue/batch-1")
	if err == nil {
		t.Fatal("expected error")
	}
}

File: src/batch.test.ts
import { describe, expect, it } from "vitest";
import { Batch, type GitOperator, type BatchPR } from "./batch.js";

function newMockGit(): GitOperator & {
  branches: string[];
  merges: string[];
  pushed: string[];
  ffRef: string;
  deleted: string[];
  conflictOn: string;
  failOn: string;
} {
  const mock = {
    branches: [] as string[],
    merges: [] as string[],
    pushed: [] as string[],
    ffRef: "",
    deleted: [] as string[],
    conflictOn: "",
    failOn: "",

    async createBranchFromRef(branch: string, _baseRef: string) {
      if (mock.failOn === "createBranchFromRef")
        throw new Error("mock error");
      mock.branches.push(branch);
    },

    async mergeBranch(
      _branch: string,
      sourceRef: string,
      _commitMsg: string,
    ) {
      if (mock.failOn === "mergeBranch") throw new Error("mock error");
      if (sourceRef === mock.conflictOn) return false;
      mock.merges.push(sourceRef);
      return true;
    },

    async pushBranch(branch: string) {
      if (mock.failOn === "pushBranch") throw new Error("mock error");
      mock.pushed.push(branch);
    },

    async getHeadSHA(ref: string) {
      if (mock.failOn === "getHeadSHA") throw new Error("mock error");
      return `mock-sha-${ref}`;
    },

    async fastForwardMain(ref: string) {
      if (mock.failOn === "fastForwardMain") throw new Error("mock error");
      mock.ffRef = ref;
      return "abc123";
    },

    async deleteBranch(branch: string) {
      if (mock.failOn === "deleteBranch") throw new Error("mock error");
      mock.deleted.push(branch);
    },
  };
  return mock;
}

const nop = () => {};

describe("CreateAndMerge", () => {
  it("merges all PRs successfully", async () => {
    const git = newMockGit();
    const b = new Batch(git, false, nop);
    const prs: BatchPR[] = [
      {
        number: 1,
        headRef: "feature-a",
        headSHA: "sha-a",
        title: "Add feature A",
      },
      {
        number: 2,
        headRef: "feature-b",
        headSHA: "sha-b",
        title: "Add feature B",
      },
    ];

    const result = await b.createAndMerge("test-1", prs);

    expect(result.branch).toBe("merge-queue/batch-test-1");
    expect(result.merged).toHaveLength(2);
    expect(result.conflicted).toHaveLength(0);
    expect(git.branches).toEqual(["merge-queue/batch-test-1"]);
    expect(git.merges).toHaveLength(2);
    expect(git.pushed).toHaveLength(1);
    expect(result.headSHA).toBe("mock-sha-merge-queue/batch-test-1");
  });

  it("sets headSHA after a successful push", async () => {
    const git = newMockGit();
    const b = new Batch(git, false, nop);
    const prs: BatchPR[] = [
      { number: 1, headRef: "feature-a", headSHA: "sha-a", title: "A" },
    ];

    const result = await b.createAndMerge("sha-test", prs);

    expect(result.headSHA).toBe("mock-sha-merge-queue/batch-sha-test");
  });

  it("leaves headSHA undefined and warns when getHeadSHA fails after push", async () => {
    const git = newMockGit();
    git.failOn = "getHeadSHA";
    const warned: string[] = [];
    const reporter = {
      info: () => {},
      async warn(msg: string) {
        warned.push(msg);
      },
      async withScope<T>(_prs: number[], fn: () => Promise<T>) {
        return fn();
      },
    };
    const b = new Batch(git, false, nop, reporter);
    const prs: BatchPR[] = [
      { number: 1, headRef: "feature-a", headSHA: "sha-a", title: "A" },
    ];

    const result = await b.createAndMerge("sha-fail", prs);

    expect(result.headSHA).toBeUndefined();
    expect(warned).toHaveLength(1);
    expect(warned[0]).toContain("failed to capture head SHA");
    expect(warned[0]).toContain("merge-queue/batch-sha-fail");
  });

  it("records conflicting PRs", async () => {
    const git = newMockGit();
    git.conflictOn = "sha-b";
    const b = new Batch(git, false, nop);
    const prs: BatchPR[] = [
      { number: 1, headRef: "feature-a", headSHA: "sha-a", title: "A" },
      { number: 2, headRef: "feature-b", headSHA: "sha-b", title: "B" },
      { number: 3, headRef: "feature-c", headSHA: "sha-c", title: "C" },
    ];

    const result = await b.createAndMerge("test-2", prs);

    expect(result.merged).toHaveLength(2);
    expect(result.conflicted).toHaveLength(1);
    expect(result.conflicted[0].number).toBe(2);
  });

  it("dry run does not create branches or push", async () => {
    const git = newMockGit();
    const b = new Batch(git, true, nop);
    const prs: BatchPR[] = [
      { number: 1, headRef: "feature-a", headSHA: "sha-a", title: "A" },
    ];

    const result = await b.createAndMerge("dry", prs);

    expect(result.merged).toHaveLength(1);
    expect(git.branches).toHaveLength(0);
    expect(git.pushed).toHaveLength(0);
  });

  it("cleans up branch on MergeBranch error", async () => {
    const git = newMockGit();
    git.failOn = "mergeBranch";
    const b = new Batch(git, false, nop);
    await expect(
      b.createAndMerge("err", [
        { number: 1, headRef: "f", headSHA: "sha-f", title: "T" },
      ]),
    ).rejects.toThrow("merging PR #1");
    expect(git.deleted).toContain("merge-queue/batch-err");
  });

  it("warns via Reporter when the cleanup deleteBranch also fails", async () => {
    // Two-level failure: mergeBranch throws (triggers cleanup), and
    // the deleteBranch teardown itself throws. The Reporter.warn call
    // should surface the teardown failure; the original merge error
    // still propagates as the thrown exception.
    const git: GitOperator = {
      async createBranchFromRef() {},
      async mergeBranch() {
        throw new Error("boom-merge");
      },
      async pushBranch() {},
      async fastForwardMain() {
        return "sha";
      },
      async deleteBranch() {
        throw new Error("boom-delete");
      },
    };
    const warned: Array<{ msg: string; scope: number[] }> = [];
    let scope: number[] = [];
    const reporter = {
      info: () => {},
      async warn(msg: string) {
        warned.push({ msg, scope: [...scope] });
      },
      async withScope<T>(prs: number[], fn: () => Promise<T>) {
        const prev = scope;
        scope = prs;
        try {
          return await fn();
        } finally {
          scope = prev;
        }
      },
    };
    const b = new Batch(git, false, nop, reporter);
    await expect(
      b.createAndMerge("err", [
        { number: 7, headRef: "f", headSHA: "sha-f", title: "T" },
      ]),
    ).rejects.toThrow("merging PR #7");

    expect(warned).toHaveLength(1);
    expect(warned[0].msg).toContain("failed to delete batch branch");
    expect(warned[0].msg).toContain("boom-delete");
    // The warning is scoped to the batch's PRs (7) at the moment it
    // fires — confirming Batch.createAndMerge set scope via
    // reporter.withScope before calling GitOps.
    expect(warned[0].scope).toEqual([7]);
  });

  it("defaults log to a nop when undefined and still exercises log call sites", async () => {
    const git = newMockGit();
    // Passing undefined for log must not throw when Batch internally calls
    // this.log (e.g. "Creating batch branch ..."). A real merge path forces
    // those log calls to execute.
    const b = new Batch(git, false, undefined);
    const result = await b.createAndMerge("test", [
      { number: 1, headRef: "f", headSHA: "sha-f", title: "T" },
    ]);
    expect(result.merged.map((p) => p.number)).toEqual([1]);
  });

  it("routes status through a custom Reporter (withScope + info)", async () => {
    const git = newMockGit();
    const scopes: number[][] = [];
    const infos: string[] = [];
    const customReporter = {
      info: (msg: string) => infos.push(msg),
      async warn() {},
      async withScope<T>(prs: number[], fn: () => Promise<T>) {
        scopes.push(prs);
        return fn();
      },
    };
    const logs: string[] = [];
    const b = new Batch(git, false, (m: string) => logs.push(m), customReporter);
    const result = await b.createAndMerge("test", [
      { number: 11, headRef: "f", headSHA: "sha-f", title: "T" },
    ]);
    expect(result.merged.map((p) => p.number)).toEqual([11]);
    // Reporter.withScope was invoked with the batch's PR numbers.
    expect(scopes).toEqual([[11]]);
    // Custom log received output.
    expect(logs.length).toBeGreaterThan(0);
  });

  it("propagates CreateBranchFromRef error", async () => {
    const git = newMockGit();
    git.failOn = "createBranchFromRef";
    const b = new Batch(git, false, nop);
    await expect(
      b.createAndMerge("err", [
        { number: 1, headRef: "f", headSHA: "sha-f", title: "T" },
      ]),
    ).rejects.toThrow();
  });

  it("handles empty PR list", async () => {
    const git = newMockGit();
    const b = new Batch(git, false, nop);
    const result = await b.createAndMerge("empty", []);

    expect(result.merged).toHaveLength(0);
    expect(git.pushed).toHaveLength(0);
  });
});

describe("CompleteMerge", () => {
  it("fast-forwards main and deletes branch", async () => {
    const git = newMockGit();
    const b = new Batch(git, false, nop);
    await b.completeMerge("merge-queue/batch-1");

    expect(git.ffRef).toBe("merge-queue/batch-1");
    expect(git.deleted).toEqual(["merge-queue/batch-1"]);
  });

  it("dry run does not fast-forward", async () => {
    const git = newMockGit();
    const b = new Batch(git, true, nop);
    await b.completeMerge("merge-queue/batch-1");

    expect(git.ffRef).toBe("");
  });

  it("propagates FastForwardMain error", async () => {
    const git = newMockGit();
    git.failOn = "fastForwardMain";
    const b = new Batch(git, false, nop);
    await expect(
      b.completeMerge("merge-queue/batch-1"),
    ).rejects.toThrow();
  });
});
